Upload and check

In [1]:
import pandas as pd

# Upload files

df = pd.read_csv('survey_results_public.csv', encoding='utf-8', low_memory=False)

df_questions = pd.read_csv('survey_results_schema.csv', encoding='utf-8', low_memory=False)

# Check result
print(f"upload lines: {len(df)}")
df.head()

upload lines: 49191


,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [2]:
pd.set_option('display.max_columns', None)

Task 1

In [3]:
# Count unique values ResponseID
total_respondents_unique = df['ResponseId'].nunique()

# Output results
total_respondents_unique

49191

Task 2

In [4]:
# 1. List QName for schema
schema_questions = set(df_questions['qname'])

# 2. List all df columns
data_columns = set(df.columns)

# 3. Intersection: only identical columns
relevant_questions = list(schema_questions.intersection(data_columns))

# 4. Create subset
df_subset = df[relevant_questions]

# 5. Check for complete responses
# dropna() removes rows with any NaN
respondents_with_all_answers = df_subset.dropna()

# 6. Final_count
final_count = len(respondents_with_all_answers)

# Print final_count
print(f"Respondents_with_all_answers: {final_count}")
print(f"We checked the answers to {len(relevant_questions)} questions.")

Respondents_with_all_answers: 0
We checked the answers to 126 questions.


Результат 0 респондентів є логічним для даного датасету, оскільки опитування містить 126 специфічних запитань, які залежать від попередніх відповідей, тому жоден респондент не заповнив абсолютно всі поля.

Task 3

In [5]:
# Calc mean, median, mode
mean_exp = df['WorkExp'].mean()
median_exp = df['WorkExp'].median()
mode_exp = df['WorkExp'].mode()[0] 

# Create summary table
stats_work_exp = pd.DataFrame({
    'Metric': ['Mean', 'Median', 'Mode'],
    'Value': [mean_exp, median_exp, mode_exp]
})
# Display the table
stats_work_exp

,Metric,Value
0,Mean,13.367403
1,Median,10.000000
2,Mode,10.000000


Task 4

In [6]:
# Sum all True values 'Remote'
remote_count = (df['RemoteWork'] == 'Remote').sum()
# Display result
print(f"Remote workers: {remote_count}")

Remote workers: 10931


Task 5

In [7]:
# Check for 'Python' (ignore NaNs) and create mask
python_mask = df['LanguageHaveWorkedWith'].str.contains('python', case=False, na=False)
python_users = python_mask.sum()
# Count only non-empty responses
total_ans = df['ResponseId'].nunique()

# Calculate and format percentage
pct_python = round((python_users / total_ans) * 100, 2)
print (total_ans)
print (python_users)
print(f"Python popularity: {pct_python}%")

49191
18466
Python popularity: 37.54%


Task 6

In [8]:
# Count online learners
# Column 'LearnCode' contains multiple values separated by semicolons
online_learners = df['LearnCode'].str.contains('Online Courses', na=False).sum()

# Display the final number
print(f"Number of online learners: {online_learners}")

Number of online learners: 10973


Task 7

In [9]:
# Reuse the Python mask from Task 5
python_df = df[python_mask]

# Group by filtered data
country_stats = python_df.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median']).dropna()
country_stats['mean'] = country_stats['mean'].round(2)
country_stats['median'] = country_stats['median'].round(0)
# Rename for better look
country_stats.columns = ['Mean Salary', 'Median Salary']
# Display result
country_stats.sort_values(by='Median Salary', ascending=False).head(10)

,Mean Salary,Median Salary
Country,,
Oman,390135.00,390135.0
Andorra,226103.50,226104.0
United States of America,173298.59,150000.0
Israel,135828.37,142594.0
Switzerland,156456.60,142592.0
Nomadic,120131.57,139218.0
Ireland,120523.92,116015.0
Luxembourg,116014.71,109054.0
Kyrgyzstan,106008.50,106008.0


Task 8

In [10]:
# ascending=False means from highest to lowest
top_5 = df.sort_values(by='ConvertedCompYearly', ascending=False).head(5)
# Display top 5 education levels
top_5[['ConvertedCompYearly', 'EdLevel']]

,ConvertedCompYearly,EdLevel
34267,50000000.0,"Associate degree (A.A., A.S., etc.)"
28700,33552715.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"
43143,18387548.0,"Associate degree (A.A., A.S., etc.)"
35353,15430267.0,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)"
45971,13921760.0,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)"


Task 9

In [11]:
# Create a temporary column in the main df to mark Python users
# Use the mask from Task 5
df['PythonUser'] = python_mask

# Group by Age and calculate the percentage of Python users
python_by_age = df.groupby('Age')['PythonUser'].mean() * 100

# Convert to DataFrame and reset index for a clean look
python_by_age_df = python_by_age.reset_index()
python_by_age_df.columns = ['Age Group', 'Python Popularity (%)']

# Display results
python_by_age_df

,Age Group,Python Popularity (%)
0,18-24 years old,40.000000
1,25-34 years old,36.939282
2,35-44 years old,36.719281
3,45-54 years old,38.629482
4,55-64 years old,37.242955
5,65 years or older,31.634820
6,Prefer not to say,31.216931


Task 10

In [12]:
# 1. Calculate 75th percentile for compensation
q75 = df['ConvertedCompYearly'].quantile(0.75)
# 2. Filter: high earners + remote workers
high_paid_remote = df[
    (df['ConvertedCompYearly'] >= q75) & 
    (df['RemoteWork'] == 'Remote')
]
# 3. Count top industries
# Most popular industries for top-tier remote roles
top_industries = high_paid_remote['Industry'].value_counts().head(10)
# Display result
top_industries

Industry
Software Development                          1186
Fintech                                        190
Healthcare                                     188
Other:                                         176
Internet, Telecomm or Information Services     138
Banking/Financial Services                      88
Government                                      78
Media & Advertising Services                    75
Retail and Consumer Services                    65
Transportation, or Supply Chain                 63
Name: count, dtype: int64